crates/kubelet/src/node.rs

```diff
@@ -166,6 +167,124 @@ pub async fn create_node<P: 'static + Provider + Sync + Send>(
     info!("Successfully created node '{}'", &config.node_name);
 }
 
+/// Fetch the uid of a node by name.
+pub async fn node_uid(client: &kube::Client, node_name: &str) -> anyhow::Result<String> {
+    let node_client: Api<KubeNode> = Api::all(client.clone());
+    match retry!(node_client.get(node_name).await, times: 4, log_error: |e| error!("Failed to get node to cordon: {:?}", e))
+    {
+        Ok(node) => match node.metadata.and_then(|meta| meta.uid) {
+            Some(uid) => Ok(uid),
+            None => {
+                let err = format!("Node missing metadata or uid {}.", node_name);
+                error!("{}", &err);
+                anyhow::bail!(err);
+            }
+        },
+        Err(e) => {
+            error!("Error fetching node {} id: {:?}.", node_name, e);
+            anyhow::bail!(e);
+        }
+    }
+}
+
+/// Cordons node and evicts all pods.
+pub async fn drain_node(client: &kube::Client, node_name: &str) -> anyhow::Result<()> {
+    cordon_node(client, node_name).await?;
+    evict_pods(client, node_name).await?;
+    Ok(())
+}
+
+/// Fetches list of pods on this node and deletes them.
+pub async fn evict_pods(client: &kube::Client, node_name: &str) -> anyhow::Result<()> {
+    let pod_client: Api<Pod> = Api::all(client.clone());
+    let node_selector = format!("spec.nodeName={}", node_name);
+    let params = ListParams {
+        field_selector: Some(node_selector),
+        ..Default::default()
+    };
+    let kube::api::ObjectList { items: pods, .. } = pod_client.list(&params).await?;
+
+    warn!("Evicting {} pods.", pods.len());
+
+    // TODO: Filter mirror pods,daemonsets, kube-system?
+    for pod in &pods {
+        if let Some(name) = pod.metadata.as_ref().and_then(|meta| meta.name.as_ref()) {
+            let namespace = pod
+                .metadata
+                .as_ref()
+                .and_then(|meta| meta.namespace.clone())
+                .unwrap_or_else(|| "default".to_owned());
+            match evict_pod(&client, name, &namespace).await {
+                Ok(_) => (),
+                Err(e) => {
+                    // Absorb the error and attempt to delete other pods with best effort.
+                    error!("Error evicting pod: {:?}", e)
+                }
+            }
+        } else {
+            error!("Pod with no name: {:?}", pod.metadata);
+        }
+    }
+    Ok(())
+}
+
+async fn evict_pod(client: &kube::Client, name: &str, namespace: &str) -> anyhow::Result<()> {
+    let ns_client: Api<Pod> = Api::namespaced(client.clone(), namespace);
+    let lp = ListParams::default()
+        .fields(&format!("metadata.name={}", name))
+        .fields(&format!("metadata.namespace={}", namespace));
+
+    // The delete call may return a "pending" response, we must watch for the actual delete event.
+    // TODO Timeout?
+    let mut stream = ns_client.watch(&lp, "0").await?.boxed();
```

Comment:
I'm not sure if it matters much in this PR, but I thought I'd bring it up: We might want to check if there is any performance overhead for spinning up a separate watch for each pod. If there are 100s of pods on the node, I don't know what the API Server does with the list params and what the overhead is for opening a stream for each individual pod. If it is a problem, we might want to see if there is a good way to share the same watch

Reply:
True, I'm not sure how `kubectl` [does it](https://kubernetes.io/images/docs/kubectl_drain.svg), I don't think its very sophisticated. I think it would be possible to do with a single `Stream` so I can give that a try.

Reply:
If it becomes too complex, don't sweat it, we can just leave a comment and revisit later